# Cuarto trabajo - Técnicas de Aprendizaje Estadístico


*   Diego de Oliveira Hitzges
*   Julián Úsuga Ortiz
*   Juan José Hurtado Álvarez
*   Santiago Campos Giraldo




## Objetivo
Predecir el numero de vehiculos registrados diariamente en el sistema de tránsito nacional RUNT

## Planteamiento del problema 
Para nuestro problema que es predecir el numero de vehiculos registados diariamente en el RUNT pensamos en realizar una red neuronal, ya que esta nos permite encontrar patrones en los datos 

# Carga de los datos 

In [1]:
from google.colab import files
uploaded = files.upload()

Saving registros_autos_entrenamiento - Hoja1.csv to registros_autos_entrenamiento - Hoja1.csv


In [2]:
import pandas as pd 
import time
import numpy as np
import random
import pickle

df = pd.read_csv("registros_autos_entrenamiento - Hoja1.csv") 

df.head()

Fecha  Unidades
0  1/1/2012         0
1  2/1/2012       188
2  3/1/2012       482
3  4/1/2012       927
4  5/1/2012      1159

# Preprocesamiento de los datos

Importamos y convertimos a dataframe, cambiamos el formato de la fecha de string a datetime


In [3]:
df.Fecha = pd.to_datetime(df.Fecha, format= '%d/%m/%Y' )
df.head()

Fecha  Unidades
0 2012-01-01         0
1 2012-01-02       188
2 2012-01-03       482
3 2012-01-04       927
4 2012-01-05      1159

Obtenemos el mes desde la fecha para usarlo como variable

In [4]:
def get_month(fecha):
  return fecha.month
df ['Mes'] = df.Fecha.apply(get_month)

In [5]:
df.head()

Fecha  Unidades  Mes
0 2012-01-01         0    1
1 2012-01-02       188    1
2 2012-01-03       482    1
3 2012-01-04       927    1
4 2012-01-05      1159    1

Obtenemos el día de la semana  desde la misma fecha para usarlo como variable 

In [6]:
def get_weekday(fecha):
  return fecha.day_name()
df['Día'] = df.Fecha.apply(get_weekday)

Tendremos en cuenta otra variable, si es un día de semana ordinario o si es sábado o festivo. De la fecha obtenemos si es sábado o domingo. Manualmente con un calendario tendremos que modificar los festivos

In [7]:
def tipo_dia (dia):
  a= 0
  if (dia == 'Sunday' ): return 'Domingos y festivos'
  else: return 'Semana'

df ['Tipo de día'] = df['Día'].apply(tipo_dia)
df

Fecha  Unidades  Mes        Día          Tipo de día
0    2012-01-01         0    1     Sunday  Domingos y festivos
1    2012-01-02       188    1     Monday               Semana
2    2012-01-03       482    1    Tuesday               Semana
3    2012-01-04       927    1  Wednesday               Semana
4    2012-01-05      1159    1   Thursday               Semana
...         ...       ...  ...        ...                  ...
2187 2017-12-27      2102   12  Wednesday               Semana
2188 2017-12-28      3449   12   Thursday               Semana
2189 2017-12-29      1084   12     Friday               Semana
2190 2017-12-30        18   12   Saturday               Semana
2191 2017-12-31         0   12     Sunday  Domingos y festivos

[2192 rows x 5 columns]

In [8]:
def get_day(fecha):
    return fecha.day
df ['Día del mes'] = df.Fecha.apply(get_day)

df.head()

Fecha  Unidades  Mes        Día          Tipo de día  Día del mes
0 2012-01-01         0    1     Sunday  Domingos y festivos            1
1 2012-01-02       188    1     Monday               Semana            2
2 2012-01-03       482    1    Tuesday               Semana            3
3 2012-01-04       927    1  Wednesday               Semana            4
4 2012-01-05      1159    1   Thursday               Semana            5

## Carga de los días festivos




In [9]:
## 2012
df.loc[df.Fecha ==('2012-01-09 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2012-03-19 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2012-04-05 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2012-04-06 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2012-05-01 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2012-05-21 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2012-06-11 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2012-06-18 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2012-07-02 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2012-07-20 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2012-08-07 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2012-08-20 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2012-10-15 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2012-11-05 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2012-11-12 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2012-12-08 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2012-12-25 00:00:00'), 'Tipo de día'] = "Domingos y festivos"

##2013
df.loc[df.Fecha ==('2013-01-01 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2013-01-07 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2013-03-25 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2013-03-28 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2013-03-29 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2013-05-01 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2013-05-13 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2013-06-03 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2013-06-10 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2013-07-01 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2013-07-20 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2013-08-17 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2013-08-19 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2013-10-14 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2013-11-04 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2013-11-11 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2013-12-25 00:00:00'), 'Tipo de día'] = "Domingos y festivos"

##2014
df.loc[df.Fecha ==('2014-01-01 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2014-01-06 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2014-03-24 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2014-04-17 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2014-04-18 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2014-05-01 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2014-06-02 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2014-06-23 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2014-06-30 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2014-08-07 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2014-08-18 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2014-10-13 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2014-11-03 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2014-11-17 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2014-12-08 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2014-12-25 00:00:00'), 'Tipo de día'] = "Domingos y festivos"

## 2015
df.loc[df.Fecha ==('2015-01-01 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2015-01-12 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2015-03-23 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2015-04-02 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2015-04-03 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2015-05-01 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2015-05-18 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2015-06-08 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2015-06-15 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2015-06-29 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2015-07-20 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2015-08-07 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2015-08-17 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2015-10-12 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2015-11-02 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2015-11-16 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2015-12-08 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2015-12-25 00:00:00'), 'Tipo de día'] = "Domingos y festivos"

## 2016
df.loc[df.Fecha ==('2016-01-01 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2016-01-11 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2016-03-21 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2016-03-24 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2016-03-25 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2016-05-09 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2016-05-30 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2016-06-06 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2016-07-04 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2016-07-20 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2016-08-15 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2016-10-17 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2016-11-07 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2016-11-14 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2016-12-08 00:00:00'), 'Tipo de día'] = "Domingos y festivos"

##2017
df.loc[df.Fecha ==('2017-01-09 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2017-03-20 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2017-04-13 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2017-04-14 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2017-05-01 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2017-05-29 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2017-06-19 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2017-06-26 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2017-07-03 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2017-07-20 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2017-08-07 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2017-08-21 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2017-10-16 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2017-11-06 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2017-11-13 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2017-12-08 00:00:00'), 'Tipo de día'] = "Domingos y festivos"
df.loc[df.Fecha ==('2017-12-25 00:00:00'), 'Tipo de día'] = "Domingos y festivos"


In [10]:
df.head()

Fecha  Unidades  Mes        Día          Tipo de día  Día del mes
0 2012-01-01         0    1     Sunday  Domingos y festivos            1
1 2012-01-02       188    1     Monday               Semana            2
2 2012-01-03       482    1    Tuesday               Semana            3
3 2012-01-04       927    1  Wednesday               Semana            4
4 2012-01-05      1159    1   Thursday               Semana            5

Analizaremos los días que se vendieron 0 unidades 

In [11]:
a = df[df.Unidades == 0]
a[["Tipo de día", "Unidades"]]

Tipo de día  Unidades
0     Domingos y festivos         0
7     Domingos y festivos         0
8     Domingos y festivos         0
21    Domingos y festivos         0
35    Domingos y festivos         0
...                   ...       ...
2170  Domingos y festivos         0
2177  Domingos y festivos         0
2184  Domingos y festivos         0
2185  Domingos y festivos         0
2191  Domingos y festivos         0

[272 rows x 2 columns]

Al parecer está directamente relacionado con ser sábado o festivo, como la siguiente tabla tiene la misma cantidad de filas que la anterior, sabemos que los sábados y los festivos la cantidad de carros registrados es 0

In [12]:
c = (a[a['Tipo de día'] == 'Domingos y festivos'])
c

Fecha  Unidades  Mes     Día          Tipo de día  Día del mes
0    2012-01-01         0    1  Sunday  Domingos y festivos            1
7    2012-01-08         0    1  Sunday  Domingos y festivos            8
8    2012-01-09         0    1  Monday  Domingos y festivos            9
21   2012-01-22         0    1  Sunday  Domingos y festivos           22
35   2012-02-05         0    2  Sunday  Domingos y festivos            5
...         ...       ...  ...     ...                  ...          ...
2170 2017-12-10         0   12  Sunday  Domingos y festivos           10
2177 2017-12-17         0   12  Sunday  Domingos y festivos           17
2184 2017-12-24         0   12  Sunday  Domingos y festivos           24
2185 2017-12-25         0   12  Monday  Domingos y festivos           25
2191 2017-12-31         0   12  Sunday  Domingos y festivos           31

[269 rows x 6 columns]

In [13]:
df

Fecha  Unidades  Mes        Día          Tipo de día  Día del mes
0    2012-01-01         0    1     Sunday  Domingos y festivos            1
1    2012-01-02       188    1     Monday               Semana            2
2    2012-01-03       482    1    Tuesday               Semana            3
3    2012-01-04       927    1  Wednesday               Semana            4
4    2012-01-05      1159    1   Thursday               Semana            5
...         ...       ...  ...        ...                  ...          ...
2187 2017-12-27      2102   12  Wednesday               Semana           27
2188 2017-12-28      3449   12   Thursday               Semana           28
2189 2017-12-29      1084   12     Friday               Semana           29
2190 2017-12-30        18   12   Saturday               Semana           30
2191 2017-12-31         0   12     Sunday  Domingos y festivos           31

[2192 rows x 6 columns]

Reemplazamos el dia en texto a numero en las columnas "Día de la semana" y "Día laboral"


In [14]:
weekdays_dic = {"Sunday":0, "Monday":1, "Tuesday":2, "Wednesday":3, "Thursday":4, "Friday":5, "Saturday":6}
laborday_dic = {"Domingos y festivos":0, "Semana":1}

In [16]:
df ['Día de la semana'] = df['Día'].replace(weekdays_dic)
df ['Día laboral'] = df['Tipo de día'].replace(laborday_dic)

df

Fecha  Unidades  Mes  ... Día del mes Día de la semana  Día laboral
0    2012-01-01         0    1  ...           1                0            0
1    2012-01-02       188    1  ...           2                1            1
2    2012-01-03       482    1  ...           3                2            1
3    2012-01-04       927    1  ...           4                3            1
4    2012-01-05      1159    1  ...           5                4            1
...         ...       ...  ...  ...         ...              ...          ...
2187 2017-12-27      2102   12  ...          27                3            1
2188 2017-12-28      3449   12  ...          28                4            1
2189 2017-12-29      1084   12  ...          29                5            1
2190 2017-12-30        18   12  ...          30                6            1
2191 2017-12-31         0   12  ...          31                0            0

[2192 rows x 8 columns]

Seleccionamos los campos 'Mes', 'Día del mes', 'Día de la semana', 'Día laboral', 'Unidades' para entrenar nuestro modelo

In [17]:
column_selection = df [['Mes', 'Día del mes', 'Día de la semana', 'Día laboral', 'Unidades']]
column_selection

Mes  Día del mes  Día de la semana  Día laboral  Unidades
0       1            1                 0            0         0
1       1            2                 1            1       188
2       1            3                 2            1       482
3       1            4                 3            1       927
4       1            5                 4            1      1159
...   ...          ...               ...          ...       ...
2187   12           27                 3            1      2102
2188   12           28                 4            1      3449
2189   12           29                 5            1      1084
2190   12           30                 6            1        18
2191   12           31                 0            0         0

[2192 rows x 5 columns]

Transformamos nuestros datos a un array, cada fila será un dia con sus correspondieste valor de Unidades vendidas

In [18]:
#transform panda to numpy_array
Data = column_selection.to_numpy()

#rows are samples
Data

array([[   1,    1,    0,    0,    0],
       [   1,    2,    1,    1,  188],
       [   1,    3,    2,    1,  482],
       ...,
       [  12,   29,    5,    1, 1084],
       [  12,   30,    6,    1,   18],
       [  12,   31,    0,    0,    0]])

Como podemos ver tenemos 2192 datos para entrenar a nuestro modelo

In [19]:
np.shape(Data)

(2192, 5)

In [20]:
#shuffle order of samples
np.random.seed(0)
np.random.shuffle(Data)
Data

array([[   9,   21,    5,    1, 1231],
       [   2,    2,    2,    1,  694],
       [  11,   10,    6,    1,  336],
       ...,
       [   2,    2,    0,    0,    6],
       [   4,   15,    2,    1, 1133],
       [   7,   11,    1,    1,  643]])

In [21]:
#transpose X, such that columns are now samples
Data = Data.T
Data

array([[   9,    2,   11, ...,    2,    4,    7],
       [  21,    2,   10, ...,    2,   15,   11],
       [   5,    2,    6, ...,    0,    2,    1],
       [   1,    1,    1, ...,    0,    1,    1],
       [1231,  694,  336, ...,    6, 1133,  643]])

Mezclamos los datos para que no queden en orden y sacamos la transpuesta, ahora cada columna será un dia con el ultimo valor como su valor de Unidades vendidas


\begin{bmatrix}
    \vert & \vert &  & \vert \\
    v_1 & v_2  & ...  & v_{2192}\\
    \vert & \vert & & \vert
\end{bmatrix}


In [24]:
#split into training and test set
Data_train = Data[:,:1500]
Data_test = Data[:,1500:]

In [25]:
np.shape(Data_train), np.shape(Data_test)

((5, 1500), (5, 692))

Ahora divido mis datos, mis datos para entrenamiento seran aprox. el 68% de mis datos totales, mis datos para validacion serán el 32%

In [26]:
#split X and target Variable Y = 'Unidades'
X_train = Data_train[:-1,:]
Y_train = Data_train[[-1],:]

X_test = Data_test[:-1,:]
Y_test = Data_test[[-1],:]

X_train, Y_train

(array([[ 9,  2, 11, ...,  9,  9,  3],
        [21,  2, 10, ..., 18, 14, 31],
        [ 5,  2,  6, ...,  5,  6,  5],
        [ 1,  1,  1, ...,  1,  1,  1]]),
 array([[1231,  694,  336, ..., 1034,  303, 2092]]))

Divido ahora mis datos para tener 4 variables para entrenar el modelo

In [27]:
np.shape(X_train), np.shape(Y_train), np.shape(X_test), np.shape(Y_test)

((4, 1500), (1, 1500), (4, 692), (1, 692))

In [28]:
#normalize Y as only not categorical variable

Y_mean = np.mean(Y_train, axis=1) 
Y_std = np.std(Y_train, axis=1)
Y_train =  (Y_train - Y_mean) / Y_std
Y_test =  (Y_test - Y_mean) / Y_std

Y_train

array([[ 0.79239981, -0.17076834, -0.81288043, ...,  0.43905879,
        -0.87206954,  2.33669733]])

Vamos a hacer una regresión con una red neuronal. Esta va a consistir en un Input-Layer, un Hidden-Layer y un Output-Layer .
- El Input será el Training set que acabamos de definir.
- El Hidden-Layer tendrá la dimensión 4 y como función de activación el Leaky ReLu
- El Output_Layer será de la dimensión 1 y no tendrá función de activación.

El resultado del outputlayer es la predición de ventas de carros en este día.
Para entrenar la red usamos el MSE como función de costo y el metodo del gradiente descente para actualizar los parametros.
Despues compararmos los resultados con el MSE del Test set.

Al final calculamos el R2 score con los parametros que obtuvimos.

# Preparando la Red Neuronal

In [29]:
#Dimensions
n_x = len(X_train)
m = len(X_train[0])

In [30]:
#Activation_function
def leaky_relu(Z):
	A = np.maximum(Z,Z*0.001)
	cache = Z
	return (A,cache)

In [31]:
#Derivative of leaky_relu
def leaky_relu_backwards(Z):
	def der_elementwise(Z):
		if Z < 0:
			return (0.001)
		else:
			return (1)
	return(np.vectorize(der_elementwise)(Z))

In [32]:
#Initialize parameters w and b
def init_parameters(n_h):
	W1 = np.random.randn(n_h,n_x) * 0.1
	b1 = np.zeros((n_h,1))
	W2 = np.random.randn(1,n_h) *0.01
	b2 = 0

	parameters = {"W1" : W1, "b1" : b1, "W2" : W2, "b2" : b2}

	return parameters

## Forward Propagation

In [33]:
def linear_forward(A,W,b):
	Z = np.dot(W,A) + b
	cache = (A,W,b)
	return (Z,cache)

In [34]:
def activation_forward(A_prev, W, b, activation):
	Z, linear_cache = linear_forward(A_prev,W,b)
	if activation == "leaky_relu":
		A, activation_cache = leaky_relu(Z)
	elif activation == "":
		A, activation_cache = Z, Z
	cache = (linear_cache, activation_cache)
	return(A,cache)

In [35]:
#forward-propagation
def two_layer_forward(X, parameters):
	caches = []
	
	W1 = parameters["W1"]
	b1 = parameters["b1"]
	A1, cache = activation_forward(X,W1,b1,"leaky_relu")
	caches.append(cache)

	W2 = parameters["W2"]
	b2 = parameters["b2"]
	A2, cache = activation_forward(A1,W2,b2,"")
	caches.append(cache)

	Y_hat = A2
	return(Y_hat,caches)

In [36]:
#Cost-function
def cost_func(Y,Y_hat):
	m = len(Y_hat[0])
	cost = 1/m * np.sum(np.power(Y-Y_hat,2))
	return(cost)

## Backward Propagation

In [37]:
def linear_backward(dZ,cache):
	A_prev, W, b = cache
	m = A_prev.shape[1]
	dW = 1/m* np.dot(dZ,A_prev.T)
	db = 1/m * np.sum(dZ, axis=1, keepdims = True)
	dA_prev = np.dot(W.T,dZ)
	return(dA_prev, dW, db)

In [38]:
def linear_activation_backward(dA, cache, activation):
	linear_cache, activation_cache = cache
	if activation == "leaky_relu":
		dZ = dA * leaky_relu_backwards(activation_cache)
		
	elif activation == "":
		dZ = dA * 1
	dA_prev, dW, db = linear_backward(dZ,linear_cache)
	return(dA_prev, dW, db)

In [39]:
#backward-propagation
def two_layer_backward(A2,Y, caches):
	grads = {}
	dA2 = (-2/m) * (Y-A2)
	
	dA_prev_temp, dW_temp, db_temp = linear_activation_backward(dA2, caches[1], "")
	grads["dA1"] = dA_prev_temp
	grads["dW2"] = dW_temp
	grads["db2"] = db_temp
	
	dA_prev_temp, dW_temp, db_temp = linear_activation_backward(grads["dA1"], caches[0], "leaky_relu")
	grads["dW1"] = dW_temp
	grads["db1"] = db_temp


	return grads

In [40]:
def update_parameters(parameters,grads,learning_rate):
	parameters["W1"] = parameters["W1"] - learning_rate * grads["dW1"]
	parameters["b1"] = parameters["b1"] - learning_rate * grads["db1"]
	parameters["W2"] = parameters["W2"] - learning_rate * grads["dW2"]
	parameters["b2"] = parameters["b2"] - learning_rate * grads["db2"]
	return parameters

## Parameter Storage

In [41]:
#read in former saved parameters
def read_parameters():
	f = open("parameters.txt", "rb")
	parameters = pickle.load(f)
	f.close()
	return parameters

In [42]:
#save the obtained parameters in a separate text-file
def save_parameters(parameters):
	f = open("parameters.txt", "wb")
	pickle.dump(parameters,f)
	f.close()

# Red Neuronal

El usuario puede cambiar la arquitectura de la red neuronal con el valor "n_h", que define la dimensión de la Capa Oculta "Hidden Layer".

Para empezar el entrenamiento hay dos opciones: Inicializar los parámetros aleatoriamente o usar los parámetros que ya estén guardados de un entrenamiento anterior. Para lo ultimo se utiliza el parámetro "read = True". Además, con el valor "read", podemos decidir si queremos guardar los parámetros después del entrenamiento o no.

Para el caso de que solo queramos guardarlos si logramos un resultado, podemos decir cuál es el error que deseamos obtener en el valor "coast_goal". En el momento que la red obtenga este valor, va a parar el entrenamiento automáticamente y guardar los parámetros relacionados.

Esto también permite empezar el entrenamiento con una tasa de aprendizaje "learning rate" más grande y cambiarla manualmente durante el proceso de entrenamiento.

Si en alguna iteración el costo aumenta, el usuario recibirá un mensaje informándolo de eso, diciendo que la tasa de aprendizaje "learning rate" está demasiado alta.

El usuario también puede decidir si quiere ver los costos durante el proceso de entrenamiento para controlar la convergencia o no. Esto se decide con el valor "print_cost".

In [43]:
#combine everything to the neural network
def neural_net(X, Y, n_h, learning_rate, epochs, cost_goal = 0, print_cost = False, read = False, save = False):

	#start either with former saved parameters or initialize them
  if read:
    parameters = read_parameters()
  else:
    parameters = init_parameters(n_h)

  #compare new cost to old cost to control convergence	
  cost_old = 9999999999999999999

  for epoch in range(epochs+1):
    #compute predicted labels in every iteration and store cache for backward-propagation
    Y_hat,caches = two_layer_forward(X, parameters)
	
    #show costs to control convergence
    cost = cost_func(Y,Y_hat)
    
    if print_cost == True:
      if epoch % 1000 == 0:
        print("epoch = " + str(epoch), "cost = " + str(cost))						

    if cost > cost_old:
      print("Learning Rate was too high")
    cost_old = cost
    

    #save gradients for gradient-descent
    grads = two_layer_backward(Y_hat,Y,caches)

    #stop and safe the parameters if the desired cost is obtained. The network can afterwards be called again to continue training with a different learning rate 
    if cost < cost_goal:
      save = True
      break
    #update parameters via gradient descent
    parameters = update_parameters(parameters,grads,learning_rate)
		

  #save final predictions and computed cost
  Y_hat,caches = two_layer_forward(X, parameters) 		
  final_cost = cost_func(Y,Y_hat)														
	
  #save obtained parameters in different text-file
  if save:
    save_parameters(parameters)

  return("Training Error is " + str(cost))

In [44]:
#Test learned parameters on test set
def test_model(X,Y):
	#gain learned parameters
	parameters = read_parameters()
	#run neural entwork to make prediction
	Y_hat,__ = two_layer_forward(X, parameters)
	#calculate costs
	cost = cost_func(Y,Y_hat)

	#show results
	return("Test Error is " + str(round(cost,2)))

# Entrenamiento y Validación

In [45]:
np.random.seed(0)
# X, Y, n_h, learning_rate, epochs, cost_goal = 0, print_cost = False, read = False, save = False)
neural_net(X_train, Y_train, 4, 2, 100000, cost_goal = 0.4)
neural_net(X_train, Y_train, 4, 1, 100000, cost_goal = 0.24, read = True)
neural_net(X_train, Y_train, 4, 0.5, 100000, cost_goal = 0.23, read = True)

'Training Error is 0.22999986681169843'

In [47]:
test_model(X_test,Y_test)

'Test Error is 0.21'

# Modelo de predicción para nuevos datos

In [52]:
#Prediction for new instance, given as a vector of attributes 
def label_new_data(x):
  #reshape X
  x = np.array(x)
  x = x.reshape((4,1))
  #check if it is a sunday or holiday:
  if x[3,0] == 0:
    return(0)
  #call learned parameters
  parameters = read_parameters()
	#predict label with neural network architecture
  Y_hat = two_layer_forward(x,parameters)[0]
  #print(Y_hat)
  #destandardize label
  Y_hat = Y_hat * Y_std + Y_mean
	#round label to integer
  Y_hat =  round(Y_hat[0][0])
	
  return Y_hat

In [48]:
np.set_printoptions(suppress=True)
Y_test = Y_test * Y_std + Y_mean
Y_test = Y_test[0]

In [53]:
Y_test[1:25]

array([1076.,  376., 1361.,    0., 1232., 1723., 1024.,  999.,  906.,
        989., 1098., 1113., 1941., 1052.,   34.,    0.,  845.,    0.,
       1072.,  840., 1311., 1056.,  277.,    0.])

In [54]:
# Predicción en test data
Y_test_pred = []
for i in range(0, len(Y_test)):
  Y_test_pred.append(label_new_data(X_test[:,i]))

In [55]:
from sklearn.metrics import r2_score

r2 = r2_score(Y_test, Y_test_pred)

r2

0.785125982646714

Como podemos ver, nuestra red neuronal tiene un $R^2$ de $0.785$ en los datos de test




## Prediccion para todos los datos (2012-2017)

In [56]:
main_df = df[['Fecha', 'Mes', 'Día del mes', 'Día de la semana', 'Día laboral']]
main_df

Fecha  Mes  Día del mes  Día de la semana  Día laboral
0    2012-01-01    1            1                 0            0
1    2012-01-02    1            2                 1            1
2    2012-01-03    1            3                 2            1
3    2012-01-04    1            4                 3            1
4    2012-01-05    1            5                 4            1
...         ...  ...          ...               ...          ...
2187 2017-12-27   12           27                 3            1
2188 2017-12-28   12           28                 4            1
2189 2017-12-29   12           29                 5            1
2190 2017-12-30   12           30                 6            1
2191 2017-12-31   12           31                 0            0

[2192 rows x 5 columns]

In [57]:
# Predecir datos 2012-2017
def pred(row):
  return label_new_data(row[['Mes', 'Día del mes', 'Día de la semana', 'Día laboral']])
main_df_pred = main_df.apply(pred, axis = 1)
main_df_pred

0          0
1        523
2        648
3        801
4       1019
        ... 
2187    1432
2188    1557
2189    1552
2190     769
2191       0
Length: 2192, dtype: int64

In [58]:
from sklearn.metrics import r2_score

r2 = r2_score(main_df_pred, df[["Unidades"]])

r2

0.7099877480586949

Como podemos ver, nuestra red neuronal, prediciendo todos los datos disponibles (de 2012 a 2017) tiene un $R^2$ de $0.7099$, lo que es bastante bueno

In [59]:
import warnings
warnings.filterwarnings('ignore')

df_2012_2017 = df[["Fecha"]]
df_2012_2017["Unidades_pred"] = main_df_pred
df_2012_2017

df_2012_2017.to_csv('df_2012_2017.csv')
files.download('df_2012_2017.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## Predicción para primer semestre de 2018

Cargo un archivo con una columna, la fecha

In [61]:
uploaded = files.upload()
df_2018 = pd.read_csv("2018.csv")
df_2018

Saving 2018.csv to 2018.csv


Fecha
0    01/01/18
1    02/01/18
2    03/01/18
3    04/01/18
4    05/01/18
..        ...
176  26/06/18
177  27/06/18
178  28/06/18
179  29/06/18
180  30/06/18

[181 rows x 1 columns]

Transformo mis datos del 2018 para tener el mismo formato

In [62]:
df_2018.Fecha = pd.to_datetime(df_2018.Fecha, format= '%d/%m/%y')

df_2018 ['Mes'] = df_2018.Fecha.apply(get_month)
df_2018['Día'] = df_2018.Fecha.apply(get_weekday)
df_2018 ['Tipo de día'] = df_2018['Día'].apply(tipo_dia)
df_2018 ['Día del mes'] = df_2018.Fecha.apply(get_day)

df_2018 ['Día de la semana'] = df_2018['Día'].replace(weekdays_dic)
df_2018 ['Día laboral'] = df_2018['Tipo de día'].replace(laborday_dic)


column_selection = df_2018 [['Mes', 'Día del mes', 'Día de la semana', 'Día laboral']]

data_2018 = column_selection.to_numpy()

Prediccion en fechas del 2018 

In [63]:
df_2018["Unidades"] = 0
df_2018

Fecha  Mes        Día  ... Día de la semana  Día laboral  Unidades
0   2018-01-01    1     Monday  ...                1            1         0
1   2018-01-02    1    Tuesday  ...                2            1         0
2   2018-01-03    1  Wednesday  ...                3            1         0
3   2018-01-04    1   Thursday  ...                4            1         0
4   2018-01-05    1     Friday  ...                5            1         0
..         ...  ...        ...  ...              ...          ...       ...
176 2018-06-26    6    Tuesday  ...                2            1         0
177 2018-06-27    6  Wednesday  ...                3            1         0
178 2018-06-28    6   Thursday  ...                4            1         0
179 2018-06-29    6     Friday  ...                5            1         0
180 2018-06-30    6   Saturday  ...                6            1         0

[181 rows x 8 columns]

In [64]:
for i in range(0, len(data_2018)):
  df_2018.loc[i, "Unidades"] = label_new_data(data_2018[i])

In [65]:
df_2018

Fecha  Mes        Día  ... Día de la semana  Día laboral  Unidades
0   2018-01-01    1     Monday  ...                1            1       506
1   2018-01-02    1    Tuesday  ...                2            1       630
2   2018-01-03    1  Wednesday  ...                3            1       800
3   2018-01-04    1   Thursday  ...                4            1      1018
4   2018-01-05    1     Friday  ...                5            1       945
..         ...  ...        ...  ...              ...          ...       ...
176 2018-06-26    6    Tuesday  ...                2            1      1167
177 2018-06-27    6  Wednesday  ...                3            1      1292
178 2018-06-28    6   Thursday  ...                4            1      1417
179 2018-06-29    6     Friday  ...                5            1      1404
180 2018-06-30    6   Saturday  ...                6            1       621

[181 rows x 8 columns]

In [66]:
df_2018.to_csv('df_2018.csv')
files.download('df_2018.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [70]:
#@title Bonus: Aplicación interactiva para predecir. { display-mode: "form" }
fecha = "2021-09-16" #@param {type:"date"}
festivo = "Si" #@param ["Si", "No"] {allow-input: true}
text_and_dropdown = 'value'
#'Mes', 'Día del mes', 'Día de la semana', 'Día laboral'
fecha_sel = pd.to_datetime(fecha, format= '%Y-%m-%d')

m = fecha_sel.month
dia_mes = fecha_sel.day
dia_sem = weekdays_dic[fecha_sel.day_name()]

if festivo == "Si":
  print("Prediccion de autos registrados para este dia: "+str(label_new_data([m, dia_mes, dia_sem, 0])))
else:
  print("Prediccion de autos registrados para este dia: "+str(label_new_data([m, dia_mes, dia_sem, 1])))


Prediccion de autos registrados para este dia: 0


# Bibliografia

Carter, D., 2021. Tensorflow — Neural Network Playground. [online] Playground.tensorflow.org. Available at: [https://playground.tensorflow.org/](https://playground.tensorflow.org/#activation=relu&batchSize=2&dataset=spiral&regDataset=reg-plane&learningRate=0.003&regularizationRate=0&noise=50&networkShape=2,2&seed=0.07223&showTestData=false&discretize=false&percTrainData=60&x=true&y=true&xTimesY=false&xSquared=false&ySquared=false&cosX=false&sinX=false&cosY=false&sinY=false&collectStats=false&problem=regression&initZero=false&hideText=false) [Accessed 7 September 2021].

En.wikipedia.org. 2021. Rectifier (neural networks) - Wikipedia. [online] Available at: <https://en.wikipedia.org/wiki/Rectifier_(neural_networks)#Leaky_ReLU> [Accessed 7 September 2021].

Jaokar, A., 2019. An elegant way to represent forward propagation and back propagation in a neural network. [online] Datasciencecentral.com. Available at: <https://www.datasciencecentral.com/profiles/blogs/an-elegant-way-to-represent-forward-propagation-and-back> [Accessed 7 September 2021].
